tests/kernel/timer/timer_behavior/src/main.c

```diff
@@ -64,13 +64,6 @@ ZTEST(timer_behavior, test_periodic_behavior)
 
 	k_timer_init(&periodic_timer, periodic_fn, NULL);
 
-#ifdef CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER
-	periodic_start = k_cycle_get_64();
-#else
-	periodic_start = k_cycle_get_32();
-#endif
-
-	periodic_data[0] = periodic_start;
```

Comment:
This is actually useful as it shows the initial timer offset as working. I'm not sure I agree that removing this makes sense

Reply:
I removed the first timer shot because it needs to spend some cycles between the timer interrupt isr got triggered and went into the timer handler, this spent cycles would be added into the first timer period, and we can not guarantee the first period won't beyond the bound requirement of this test after adding those cycles.

Reply:
I might need some better explanation of this behavior. If you are seeing a large time difference between the desired first timeout of the timer and the actual timeout of the timer, that seems like a bug.

k_timer_start(&timer, K_MSEC(1), K_MSEC(1)); should fire in 1msec or a hair over from the call time otherwise there's presumably a bug and the test *should* fail if its massively off.

Reply:
when we set the next timeout, it is based on last cycles count in timer isr, but between timer isr and custom defined timer handler, there will exist a drift(need to spent some cycles),

so for the first timeout, actually it is (_expected_period + drift_), and this drift relies on hardware(ideally it's 0), we cannot make sure this drift won't beyond required bound on all platforms.
but for the later timeout, like 2nd, _periodic_data[2] - periodic_data[1] = (2nd timer isr) - (1st timer isr) = (expected period)_

![image](https://user-images.githubusercontent.com/52983984/191195090-618147b8-63ba-44cd-bb58-a2bd97ae5ac7.png)

Reply:
This is really confusing behavior. If the test is failing only on the first timeout where the timer fires over 100us later or earlier than expected including the tick rounding that seems like a bug in some way to me.

Reply:
I kept the first timeout, because I found the drift won't beyond 100us on most platforms, but I added a new config TEST_TIMER_PERIOD_MAX_DRIFT_PERCENT for users to set expected percentage.